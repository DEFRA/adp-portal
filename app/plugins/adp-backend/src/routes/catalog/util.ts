export function createTransformerTitle(title: string, alias?: string) {
  const titleValue = alias ? `${title} ` + `(${alias})` : title;
  return titleValue;
}
export const ARMS_LENGTH_BODY_ID_ANNOTATION =
  'adp.defra.gov.uk/arms-length-body-id';
export const DELIVERY_PROGRAMME_ID_ANNOTATION =
  'adp.defra.gov.uk/delivery-programme-id';
export const DELIVERY_PROJECT_ID_ANNOTATION =
  'adp.defra.gov.uk/delivery-project-id';
