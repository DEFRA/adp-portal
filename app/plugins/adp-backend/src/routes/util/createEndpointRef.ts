import {
  type ServiceRef,
  createServiceFactory,
  createServiceRef,
} from '@backstage/backend-plugin-api';
import type { Request, RequestHandler } from 'express';
import { randomUUID } from 'node:crypto';
import type { HandlerResult } from './HandlerResult';
import type { ServiceRefsToInstances } from './ServiceRefsToInstances';
import { routerResultsRef, type RouterResults } from './routerResultsRef';

type Endpoint<
  P,
  ResBody,
  ReqBody,
  ReqQuery,
  Locals extends Record<string, any>,
> = (
  request: Request<P, ResBody, ReqBody, ReqQuery, Locals>,
) => HandlerResult | PromiseLike<HandlerResult>;

interface EndpointOptions<
  Dependencies extends Record<string, ServiceRef<unknown>>,
  P,
  ResBody,
  ReqBody,
  ReqQuery,
  Locals extends Record<string, any>,
> {
  deps: Dependencies;
  factory(options: {
    deps: ServiceRefsToInstances<Dependencies>;
    responses: RouterResults;
  }):
    | Endpoint<P, ResBody, ReqBody, ReqQuery, Locals>
    | PromiseLike<Endpoint<P, ResBody, ReqBody, ReqQuery, Locals>>;
}

export function createEndpointRef<
  P,
  ResBody,
  ReqBody,
  ReqQuery,
  Locals extends Record<string, any>,
  Dependencies extends Record<string, ServiceRef<unknown>>,
>(
  options: EndpointOptions<Dependencies, P, ResBody, ReqBody, ReqQuery, Locals>,
) {
  return createServiceRef<
    RequestHandler<P, ResBody, ReqBody, ReqQuery, Locals>
  >({
    id: `adp.router.endpoint.${randomUUID()}`,
    scope: 'plugin',
    defaultFactory(service) {
      const routerResultsKey = randomUUID() as 'routerResults';
      return Promise.resolve(
        createServiceFactory<
          RequestHandler<P, ResBody, ReqBody, ReqQuery, Locals>,
          RequestHandler<P, ResBody, ReqBody, ReqQuery, Locals>,
          any
        >({
          service,
          deps: { ...options.deps, [routerResultsKey]: routerResultsRef },
          async factory({ [routerResultsKey]: responses, ...deps }) {
            const handler = await options.factory({
              deps: deps as Dependencies,
              responses,
            });
            return voidify(async (req, res, next) => {
              try {
                const result = await handler(req);
                await result.writeTo(res);
              } catch (err) {
                next(err);
              }
            });
          },
        }),
      );
    },
  });
}

function voidify<This, Args extends unknown[]>(
  fn: (this: This, ...args: Args) => unknown,
): (this: This, ...args: Args) => void {
  return function voided(...args) {
    void fn.call(this, ...args);
  };
}
