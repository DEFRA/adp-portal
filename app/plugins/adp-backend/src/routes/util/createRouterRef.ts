import {
  type ServiceRef,
  createServiceFactory,
  createServiceRef,
} from '@backstage/backend-plugin-api';
import type { Router } from 'express';
import { randomUUID } from 'node:crypto';
import { routerFactoryRef } from './routerFactoryRef';
import { type RouterResults, routerResultsRef } from './routerResultsRef';
import type { ServiceRefsToInstances } from './ServiceRefsToInstances';

interface RouterFactoryOptions<
  Dependencies extends Record<string, ServiceRef<unknown, 'plugin' | 'root'>>,
> {
  deps: Dependencies;
  factory(options: {
    router: Router;
    deps: ServiceRefsToInstances<Dependencies>;
    responses: RouterResults;
  }): void | Promise<void>;
}

export function createRouterRef<
  Dependencies extends Record<string, ServiceRef<unknown, 'plugin' | 'root'>>,
>(options: RouterFactoryOptions<Dependencies>) {
  return createServiceRef<Router>({
    id: `adp.router.${randomUUID()}`,
    scope: 'plugin',
    defaultFactory(service) {
      return Promise.resolve(createRouterFactory(service, options));
    },
  });
}
function createRouterFactory<
  Dependencies extends Record<string, ServiceRef<unknown, 'plugin' | 'root'>>,
>(
  service: ServiceRef<Router, 'plugin'>,
  options: RouterFactoryOptions<Dependencies>,
) {
  const routerFactoryKey = randomUUID() as 'routerFactory';
  const routerResultsKey = randomUUID() as 'routerResults';
  return createServiceFactory<Router, Router, any>({
    service,
    deps: {
      ...options.deps,
      [routerFactoryKey]: routerFactoryRef,
      [routerResultsKey]: routerResultsRef,
    },
    async factory({
      [routerFactoryKey]: routerFactory,
      [routerResultsKey]: routerResults,
      ...deps
    }) {
      const router = routerFactory();
      await options.factory({
        router,
        deps: deps as Dependencies,
        responses: routerResults,
      });
      return router;
    },
  });
}
