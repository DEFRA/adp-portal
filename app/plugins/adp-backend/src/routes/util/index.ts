export * from './createEndpointRef';
export * from './createRouterRef';
export * from './HandlerResult';
export * from './routerFactoryRef';
export * from './routerResultsRef';
