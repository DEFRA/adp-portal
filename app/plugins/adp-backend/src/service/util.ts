import { Response } from 'express';
import { ValidationErrorMapping } from '@internal/plugin-adp-common';
import { InputError } from '@backstage/errors';
import { z } from 'zod';

export function respond<Request, Success, Error extends string>(
  request: Request,
  response: Response,
  message:
    | { success: true; value: Success }
    | { success: false; errors: Error[] },
  mapping: ValidationErrorMapping<Request, Error>,
  statuses?: { ok?: number; badRequest?: number },
) {
  if (!message.success) {
    response.status(statuses?.badRequest ?? 400).json({
      errors: message.errors.map(e => mapping[e](request)),
    });
  } else if (message.value === undefined) {
    response.status(statuses?.ok ?? 204).end();
  } else {
    response.status(statuses?.ok ?? 200).json(message.value);
  }
}

export function createParser<T>(schema: z.ZodType<T>) {
  return (body: unknown) => {
    const result = schema.safeParse(body);
    if (result.success) return result.data;
    throw new InputError(result.error.message);
  };
}

export type SafeResult<Success, Error extends PropertyKey> =
  | { success: true; value: Success }
  | { success: false; errors: Array<Error | 'unknown'> };

export const emptyUUID = '00000000-0000-0000-0000-000000000000';
export type UUID = `${string}-${string}-${string}-${string}-${string}`;

export function isUUID(value: string): value is UUID {
  return /^[\da-f]{8}-[\da-f]{4}-[\da-f]{4}-[\da-f]{4}-[\da-f]{12}$/i.test(
    value,
  );
}

export async function checkMany<
  T extends Record<string, boolean | Promise<boolean>>,
>(checks: T): Promise<SafeResult<undefined, keyof T>> {
  const results: Array<[keyof T, boolean]> = await Promise.all(
    Object.entries(checks).map(async e => [e[0], await e[1]] as const),
  );
  const failed = results.filter(x => x[1]);
  if (failed.length === 0) return { success: true, value: undefined };

  return { success: false, errors: failed.map(x => x[0]) };
}

export function containsAnyValue(obj: object) {
  return Object.entries(obj).some(e => e[1] !== undefined);
}
