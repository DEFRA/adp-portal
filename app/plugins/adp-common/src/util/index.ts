export * from './annotations';
export * from './createName';
export * from './normalizeUsername';
