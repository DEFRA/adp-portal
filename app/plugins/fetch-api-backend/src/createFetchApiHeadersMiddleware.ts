import type { Fetch } from './Fetch';
import type { FetchApiMiddleware } from './FetchApiMiddleware';

export type HeaderValue =
  | string
  | undefined
  | ((
      ...args: Parameters<Fetch>
    ) => string | undefined | Promise<string | undefined>);
export type HeadersOptions =
  | Iterable<readonly [string, HeaderValue]>
  | Readonly<Record<string, HeaderValue>>;

export function createFetchApiHeadersMiddleware(
  headers: HeadersOptions,
): FetchApiMiddleware {
  const headersArr =
    Symbol.iterator in headers ? [...headers] : Object.entries(headers);
  return fetch =>
    async function headersMiddleware(input, init) {
      if (typeof input === 'string' || !('headers' in input)) {
        const realInit = init ?? {};
        realInit.headers ??= [];
        await applyHeaders(
          headersArr,
          getAdapter(realInit.headers),
          input,
          realInit,
        );
      } else {
        await applyHeaders(headersArr, input.headers, input, init);
      }
      return fetch(input, init);
    };
}

async function applyHeaders(
  options: readonly (readonly [string, HeaderValue])[],
  headers: HeadersAdapter,
  ...[input, init]: Parameters<Fetch>
) {
  for (const [header, valueOrFactory] of options) {
    if (headers.has(header)) continue;
    const value =
      typeof valueOrFactory === 'function'
        ? await valueOrFactory(input, init)
        : valueOrFactory;
    if (value) headers.append(header, value);
  }
}

type HeadersAdapter = Pick<Headers, 'has' | 'append'>;

function getAdapter(headers: HeadersInit): HeadersAdapter {
  // class Headers { ... }
  if (isLikeHeadersClass(headers)) return headers;
  // Array<[string, string]>
  if (Array.isArray(headers)) {
    return {
      has(header) {
        const lowerHeader = header;
        return headers.some(h => h[0].toLowerCase() === lowerHeader);
      },
      append(header, value) {
        headers.push([header, value]);
      },
    };
  }
  // Record<string, string>
  return {
    has(header) {
      const lowerHeader = header;
      return Object.keys(headers).some(h => h.toLowerCase() === lowerHeader);
    },
    append(header, value) {
      const lowerHeader = header;
      const key =
        Object.keys(headers).find(h => h.toLowerCase() === lowerHeader) ??
        header;
      const currentValue = headers[key];
      headers[key] = currentValue ? `${currentValue} ${value}` : value;
    },
  };
}

function isLikeHeadersClass(
  value: unknown,
): value is { has: Function; append: Function } {
  return (
    typeof value === 'object' &&
    value !== null &&
    'append' in value &&
    typeof value.append === 'function' &&
    'has' in value &&
    typeof value.has === 'function'
  );
}
