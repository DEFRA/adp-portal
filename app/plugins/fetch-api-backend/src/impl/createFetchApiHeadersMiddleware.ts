import type { Fetch, FetchApiMiddleware } from '../types';

export type HeaderValue =
  | string
  | undefined
  | ((
      ...args: Parameters<Fetch>
    ) => string | undefined | Promise<string | undefined>);
export type HeadersOptions =
  | Iterable<readonly [string, HeaderValue]>
  | Readonly<Record<string, HeaderValue>>;

export function createFetchApiHeadersMiddleware(
  headers: HeadersOptions,
): FetchApiMiddleware {
  const headersArr =
    Symbol.iterator in headers ? [...headers] : Object.entries(headers);
  return fetch =>
    async function headersMiddleware(input, init) {
      if (typeof input !== 'string' && 'headers' in input) {
        await applyHeaders(headersArr, input.headers, input, init);
        return fetch(input, init);
      }

      const realInit = init ?? {};
      realInit.headers ??= [];
      const adapter = getAdapter(realInit.headers);
      if (await applyHeaders(headersArr, adapter, input, realInit))
        return fetch(input, realInit);
      return fetch(input, init);
    };
}

async function applyHeaders(
  options: readonly (readonly [string, HeaderValue])[],
  headers: HeadersAdapter,
  ...[input, init]: Parameters<Fetch>
) {
  let result = false;
  for (const [header, valueOrFactory] of options) {
    if (headers.has(header)) continue;
    const value =
      typeof valueOrFactory === 'function'
        ? await valueOrFactory(input, init)
        : valueOrFactory;
    if (value) {
      headers.append(header, value);
      result = true;
    }
  }
  return result;
}

type HeadersAdapter = Pick<Headers, 'has' | 'append'>;

function getAdapter(headers: HeadersInit): HeadersAdapter {
  // class Headers { ... }
  if (isLikeHeadersClass(headers)) return headers;
  // Array<[string, string]>
  if (Array.isArray(headers)) {
    return {
      has(header) {
        const lowerHeader = header.toLowerCase();
        return headers.some(h => h[0].toLowerCase() === lowerHeader);
      },
      append(header, value) {
        headers.push([header, value]);
      },
    };
  }
  // Record<string, string>
  return {
    has(header) {
      const lowerHeader = header.toLowerCase();
      return Object.keys(headers).some(h => h.toLowerCase() === lowerHeader);
    },
    append(header, value) {
      headers[header] = value;
    },
  };
}

function isLikeHeadersClass(
  value: unknown,
): value is { has: Function; append: Function } {
  return (
    typeof value === 'object' &&
    value !== null &&
    'append' in value &&
    typeof value.append === 'function' &&
    'has' in value &&
    typeof value.has === 'function'
  );
}
