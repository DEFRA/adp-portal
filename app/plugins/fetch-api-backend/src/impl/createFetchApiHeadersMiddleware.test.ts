import type { Fetch } from '../types';
import { createFetchApiHeadersMiddleware } from './createFetchApiHeadersMiddleware';
import util from 'node:util';

expect.addSnapshotSerializer({
  serialize(val, _, indentation) {
    if (!(val instanceof Request))
      throw new Error('Expected value to be a request');

    const indent = indentation;
    return util
      .inspect(val)
      .split('\n')
      .map((line, i) => (i > 0 ? indent + line : line))
      .join('\n');
  },
  test(arg0) {
    return arg0 instanceof Request;
  },
});

describe('createFetchApiHeadersMiddleware', () => {
  it.each<TestCase>([
    {
      name: 'a string input without init, but no headers',
      input: 'https://test.com',
      init: undefined,
      headers: {},
    },
    {
      name: 'a string input with an empty init, but no headers',
      input: 'https://test.com',
      init: {},
      headers: {},
    },
    {
      name: 'a string input without init',
      input: 'https://test.com',
      init: undefined,
      headers: {
        header1: 'abc',
        header2: 'def',
        header3: undefined,
        header4: () => '123',
        header5: () => undefined,
      },
    },
    {
      name: 'a string input with an empty init',
      input: 'https://test.com',
      init: {},
      headers: {
        header1: 'abc',
        header2: 'def',
        header3: undefined,
        header4: () => '123',
        header5: () => undefined,
      },
    },
    {
      name: 'a string input with an init with header array',
      input: 'https://test.com',
      init: { headers: [['Header1', 'xyz']] },
      headers: {
        header1: 'abc',
        header2: 'def',
        header3: undefined,
        header4: () => '123',
        header5: () => undefined,
      },
    },
    {
      name: 'a string input with an init with header object',
      input: 'https://test.com',
      init: { headers: { Header1: 'xyz' } },
      headers: {
        header1: 'abc',
        header2: 'def',
        header3: undefined,
        header4: () => '123',
        header5: () => undefined,
      },
    },
    {
      name: 'a string input with an init with headers instance',
      input: 'https://test.com',
      init: { headers: new Headers({ Header1: 'xyz' }) },
      headers: {
        header1: 'abc',
        header2: 'def',
        header3: undefined,
        header4: () => '123',
        header5: () => undefined,
      },
    },
    {
      name: 'a string input without init, but no headers',
      input: new URL('https://test.com'),
      init: undefined,
      headers: {},
    },
    {
      name: 'a string input with an empty init, but no headers',
      input: new URL('https://test.com'),
      init: {},
      headers: {},
    },
    {
      name: 'a url input without init',
      input: new URL('https://test.com'),
      init: undefined,
      headers: {
        header1: 'abc',
        header2: 'def',
        header3: undefined,
        header4: () => '123',
        header5: () => undefined,
      },
    },
    {
      name: 'a url input with an empty init',
      input: new URL('https://test.com'),
      init: {},
      headers: {
        header1: 'abc',
        header2: 'def',
        header3: undefined,
        header4: () => '123',
        header5: () => undefined,
      },
    },
    {
      name: 'a url input with an init with header array',
      input: new URL('https://test.com'),
      init: { headers: [['Header1', 'xyz']] },
      headers: {
        header1: 'abc',
        header2: 'def',
        header3: undefined,
        header4: () => '123',
        header5: () => undefined,
      },
    },
    {
      name: 'a url input with an init with header object',
      input: new URL('https://test.com'),
      init: { headers: { Header1: 'xyz' } },
      headers: {
        header1: 'abc',
        header2: 'def',
        header3: undefined,
        header4: () => '123',
        header5: () => undefined,
      },
    },
    {
      name: 'a url input with an init with headers instance',
      input: new URL('https://test.com'),
      init: { headers: new Headers({ Header1: 'xyz' }) },
      headers: {
        header1: 'abc',
        header2: 'def',
        header3: undefined,
        header4: () => '123',
        header5: () => undefined,
      },
    },
    {
      name: 'a request input',
      input: new Request('https://test.com', { headers: { Header1: 'xyz' } }),
      init: undefined,
      headers: {
        header1: 'abc',
        header2: 'def',
        header3: undefined,
        header4: () => '123',
        header5: () => undefined,
      },
    },
    {
      name: 'a request input without headers',
      input: new Request('https://test.com', { headers: { Header1: 'xyz' } }),
      init: undefined,
      headers: {},
    },
  ])('Should work with $name', async ({ headers, input, init }) => {
    // arrange
    const fetch: jest.MockedFn<Fetch> = jest.fn();
    const sut = createFetchApiHeadersMiddleware(headers)(fetch);
    const response = new Response();

    fetch.mockResolvedValueOnce(response);

    // act
    const actual = await sut(input, init);

    // assert
    expect(actual).toBe(response);
    expect(fetch).toHaveBeenCalledTimes(1);
    expect(fetch.mock.calls[0]).toMatchSnapshot();
  });
});

interface TestCase {
  readonly name: string;
  readonly input: Parameters<Fetch>[0];
  readonly init: Parameters<Fetch>[1];
  readonly headers: Parameters<typeof createFetchApiHeadersMiddleware>[0];
}
