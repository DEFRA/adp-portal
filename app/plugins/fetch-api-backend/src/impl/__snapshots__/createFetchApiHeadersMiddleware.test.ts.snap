// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createFetchApiHeadersMiddleware Should work with a request input 1`] = `
[
  Request {
    "body": null,
    "bodyUsed": false,
    "cache": "default",
    "credentials": "same-origin",
    "destination": "",
    "headers": Headers {
      "header1": "xyz",
      "header2": "def",
      "header4": "123",
    },
    "integrity": "",
    "keepalive": false,
    "method": "GET",
    "mode": "cors",
    "redirect": "follow",
    "referrer": "about:client",
    "referrerPolicy": "",
    "signal": AbortSignal {
      "aborted": false,
      "reason": undefined,
    },
    "url": "https://test.com/",
  },
  undefined,
]
`;

exports[`createFetchApiHeadersMiddleware Should work with a request input without headers 1`] = `
[
  Request {
    "body": null,
    "bodyUsed": false,
    "cache": "default",
    "credentials": "same-origin",
    "destination": "",
    "headers": Headers {
      "header1": "xyz",
    },
    "integrity": "",
    "keepalive": false,
    "method": "GET",
    "mode": "cors",
    "redirect": "follow",
    "referrer": "about:client",
    "referrerPolicy": "",
    "signal": AbortSignal {
      "aborted": false,
      "reason": undefined,
    },
    "url": "https://test.com/",
  },
  undefined,
]
`;

exports[`createFetchApiHeadersMiddleware Should work with a string input with an empty init 1`] = `
[
  "https://test.com",
  {
    "headers": [
      [
        "header1",
        "abc",
      ],
      [
        "header2",
        "def",
      ],
      [
        "header4",
        "123",
      ],
    ],
  },
]
`;

exports[`createFetchApiHeadersMiddleware Should work with a string input with an empty init, but no headers 1`] = `
[
  "https://test.com",
  {
    "headers": [],
  },
]
`;

exports[`createFetchApiHeadersMiddleware Should work with a string input with an empty init, but no headers 2`] = `
[
  "https://test.com/",
  {
    "headers": [],
  },
]
`;

exports[`createFetchApiHeadersMiddleware Should work with a string input with an init with header array 1`] = `
[
  "https://test.com",
  {
    "headers": [
      [
        "Header1",
        "xyz",
      ],
      [
        "header2",
        "def",
      ],
      [
        "header4",
        "123",
      ],
    ],
  },
]
`;

exports[`createFetchApiHeadersMiddleware Should work with a string input with an init with header object 1`] = `
[
  "https://test.com",
  {
    "headers": {
      "Header1": "xyz",
      "header2": "def",
      "header4": "123",
    },
  },
]
`;

exports[`createFetchApiHeadersMiddleware Should work with a string input with an init with headers instance 1`] = `
[
  "https://test.com",
  {
    "headers": Headers {
      "header1": "xyz",
      "header2": "def",
      "header4": "123",
    },
  },
]
`;

exports[`createFetchApiHeadersMiddleware Should work with a string input without init 1`] = `
[
  "https://test.com",
  {
    "headers": [
      [
        "header1",
        "abc",
      ],
      [
        "header2",
        "def",
      ],
      [
        "header4",
        "123",
      ],
    ],
  },
]
`;

exports[`createFetchApiHeadersMiddleware Should work with a string input without init, but no headers 1`] = `
[
  "https://test.com",
  undefined,
]
`;

exports[`createFetchApiHeadersMiddleware Should work with a string input without init, but no headers 2`] = `
[
  "https://test.com/",
  undefined,
]
`;

exports[`createFetchApiHeadersMiddleware Should work with a url input with an empty init 1`] = `
[
  "https://test.com/",
  {
    "headers": [
      [
        "header1",
        "abc",
      ],
      [
        "header2",
        "def",
      ],
      [
        "header4",
        "123",
      ],
    ],
  },
]
`;

exports[`createFetchApiHeadersMiddleware Should work with a url input with an init with header array 1`] = `
[
  "https://test.com/",
  {
    "headers": [
      [
        "Header1",
        "xyz",
      ],
      [
        "header2",
        "def",
      ],
      [
        "header4",
        "123",
      ],
    ],
  },
]
`;

exports[`createFetchApiHeadersMiddleware Should work with a url input with an init with header object 1`] = `
[
  "https://test.com/",
  {
    "headers": {
      "Header1": "xyz",
      "header2": "def",
      "header4": "123",
    },
  },
]
`;

exports[`createFetchApiHeadersMiddleware Should work with a url input with an init with headers instance 1`] = `
[
  "https://test.com/",
  {
    "headers": Headers {
      "header1": "xyz",
      "header2": "def",
      "header4": "123",
    },
  },
]
`;

exports[`createFetchApiHeadersMiddleware Should work with a url input without init 1`] = `
[
  "https://test.com/",
  {
    "headers": [
      [
        "header1",
        "abc",
      ],
      [
        "header2",
        "def",
      ],
      [
        "header4",
        "123",
      ],
    ],
  },
]
`;
